### PR TITLE
Add test for thread pool concurrent initialization

### PR DESCRIPTION
**Pull request by kouvel**
Depends on dotnet/coreclr#10869
Fixes dotnet/coreclr#10521

### CODE REVIEW

**Comment by kouvel** [comment]
@stephentoub @gkhanna79

File: src/System.Threading.ThreadPool/tests/ThreadPoolTests.cs [hunk 1]
```diff
@@ -14,6 +14,43 @@ public static class ThreadPoolTests
         private const int ExpectedTimeoutMilliseconds = ThreadTestHelpers.ExpectedTimeoutMilliseconds;
         private const int MaxPossibleThreadCount = 0x7fff;
 
+        static ThreadPoolTests()
+        {
+            // Run the following tests before any others
+            ConcurrentInitializeTest();
+        }
+
+        [Fact]
```
Comment (stephentoub) [inline src/System.Threading.ThreadPool/tests/ThreadPoolTests.cs]: Does this need to be marked [Fact]?  As it, it seems like it'll be invoked twice, once from the static cctor, and once as its own dedicated test.

File: src/System.Threading.ThreadPool/tests/ThreadPoolTests.cs [hunk 2]
```diff
@@ -14,6 +14,43 @@ public static class ThreadPoolTests
...
+                t.IsBackground = true;
+                t.Start();
+                threadReady.CheckedWait();
+            }
+
+            continueThreads.Set();
+
+            foreach (Action waitForThread in waitForThreadArray)
+            {
+                waitForThread();
```
Comment (stephentoub) [inline src/System.Threading.ThreadPool/tests/ThreadPoolTests.cs]: Doing synchronization in a static cctor can easily lead to deadlock.  May not be an issue here, but something to keep in mind.

File: src/System.Threading.ThreadPool/tests/ThreadPoolTests.cs [hunk 3]
```diff
@@ -14,6 +14,43 @@ public static class ThreadPoolTests
...
+            var waitForThreadArray = new Action[processorCount];
+            for (int i = 0; i < processorCount; ++i)
+            {
+                var t = ThreadTestHelpers.CreateGuardedThread(out waitForThreadArray[i], threadMain);
+                t.IsBackground = true;
+                t.Start();
+                threadReady.CheckedWait();
+            }
+
+            continueThreads.Set();
```
Comment (stephentoub) [inline src/System.Threading.ThreadPool/tests/ThreadPoolTests.cs]: Are you trying to test making all of the threads wait until they've all arrived and then releasing them all?  That could be done more simply in a variety of ways, such as creating a CountdownEvent that they all Signal and then Wait on, or creating a Barrier that they all SignalAndWait on.

File: src/System.Threading.ThreadPool/tests/ThreadPoolTests.cs (hunk 1 above)
Comment (kouvel) [inline src/System.Threading.ThreadPool/tests/ThreadPoolTests.cs]: It's not necessary, but the invocation from the static constructor is only a workaround. It looks like xunit has test priorities to have tests run in a particular order but our version does not seem to support it. I would prefer to use that when it's available, for now the extra test invocation shouldn't hurt.

File: src/System.Threading.ThreadPool/tests/ThreadPoolTests.cs (hunk 2 above)
Comment (kouvel) [inline src/System.Threading.ThreadPool/tests/ThreadPoolTests.cs]: All of the waits are timed, so in the worst case it should at least throw an exception

File: src/System.Threading.ThreadPool/tests/ThreadPoolTests.cs (hunk 3 above)
Comment (kouvel) [inline src/System.Threading.ThreadPool/tests/ThreadPoolTests.cs]: Good point, I'll look into using a barrier

**Comment by kouvel** [comment]
Barrier wasn't yielding a failure on debug build, used CountdownEvent instead

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Pull request merged**